File: 144-Binary-Tree-Preorder-Traversal/BinaryTreePreorderTraversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var preorderTraversal = function(root) {
    var res = []
    var cur = root
    var prev
    while (cur) {
        if (!cur.left) {
            res.push(cur.val)
            cur = cur.right
            continue
        }
        prev = cur.left
        while (prev.right && prev.right!=cur) {
            prev = prev.right
        }
        if (!prev.right) {
            res.push(cur.val)
            prev.right = cur
            cur = cur.left
        } else {
            prev.right = null
            cur = cur.right
        }
    }
    return res
};

File: 144-Binary-Tree-Preorder-Traversal/BinaryTreePreorderTraversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func preorderTraversal(root *TreeNode) []int {
	res := make([]int, 0)
	cur := root
	for cur != nil {
		if cur.Left == nil {
			res = append(res, cur.Val)
			cur = cur.Right
			continue
		}
		prev := cur.Left
		for prev.Right != nil && prev.Right != cur {
			prev = prev.Right
		}
		if prev.Right == nil {
			res = append(res, cur.Val)
			prev.Right, cur = cur, cur.Left
		} else {
			prev.Right, cur = nil, cur.Right
		}
	}
	return res
}
